libs/libvtrutil/src/vtr_ragged_matrix.h

```diff
@@ -0,0 +1,188 @@
+    }
+
+  public: //Accessors
+    //Iterators to *all* elements
+    auto begin() { return data_.begin(); }
+    auto end() { return data_.end(); }
+    auto begin() const { return data_.begin(); }
+    auto end() const { return data_.end(); }
+
+    size_t size() const {
+        return data_.size() - 1; //-1 for sentinel
+    }
+
+    //Indexing operators for the first dimension
+    ProxyRow<T> operator[](Index0 i) {
+        int idx = size_t(i);
+        return ProxyRow<T>(&data_[first_elem_[idx]],
+                           &data_[first_elem_[idx + 1]]);
+    }
+
+    ProxyRow<const T> operator[](Index0 i) const {
+        int idx = size_t(i);
+        return ProxyRow<const T>(&data_[first_elem_[idx]],
+                                 &data_[first_elem_[idx + 1]]);
+    }
+
+    void clear() {
+        data_.clear();
+        first_elem_.clear();
+    }
+
+  public: //Types
+    //Proxy class used to represent a 'row' in the matrix
+    template<typename U>
+    class ProxyRow {
+      public:
+        ProxyRow(U* first, U* last)
+            : first_(first)
+            , last_(last) {}
+
+        U* begin() { return first_; }
+        U* end() { return last_; }
+
+        const U* begin() const { return first_; }
+        const U* end() const { return last_; }
+
+        size_t size() const { return last_ - first_; }
+
+        U& operator[](Index1 j) {
+            VTR_ASSERT_SAFE(size_t(j) < size());
+            return first_[size_t(j)];
+        }
+
+        const U& operator[](Index1 j) const {
+            VTR_ASSERT_SAFE(size_t(j) < size());
+            return first_[size_t(j)];
+        }
+
+        U* data() {
+            return first_;
+        }
+
+        U* data() const {
+            return first_;
+        }
+
+      private:
+        U* first_;
+        U* last_;
+    };
+
+  private:
+    //Iterator for constructing FlatRaggedMatrix which uses a callback to determine
+    //row lengths.
+    template<class Callback>
+    class RowLengthIterator : public std::iterator<std::random_access_iterator_tag, size_t> {
+      public:
+        RowLengthIterator(size_t irow, Callback& callback)
+            : irow_(irow)
+            , callback_(callback) {}
+
+        RowLengthIterator& operator++() {
+            ++irow_;
+            return *this;
+        }
+
+        bool operator==(const RowLengthIterator& other) {
+            return irow_ == other.irow_;
+        }
+
+        bool operator!=(const RowLengthIterator& other) {
+            return !(*this == other);
+        }
+
+        int operator-(const RowLengthIterator& other) {
+            return irow_ - other.irow_;
+        }
+
+        size_t operator*() {
+            //Call the callback to get the row length
+            return callback_(Index0(irow_));
+        }
+
+      private:
+        size_t irow_;
+        Callback& callback_;
+    };
+
+  private:
+    std::vector<T> data_;
+    std::vector<int> first_elem_;
```

Comment:
Consider make a method to sanity check that first_elem_ is sorted?

Reply:
The structure is enforced by the constructor and there is no method to modify it after construction, so no need to check for sorting.